basic docs for nested column query functions (#12922)
apache/druid: docs/querying/virtual-columns.md

|--------|-----------|---------|

> |name|The name of the virtual column.|yes|

Reviewer: This table should include the `type` property as required. Description can be something like "Type of the virtual column. Set to `expression` for an expression virtual column."

Reviewer: Ditto for the `nested-field` virtual column and `mv-filtered` virtual column

|expression|An [expression](../misc/math-expr.md) that takes a row as input and outputs a value for the virtual column.|yes|
|outputType|The expression's output will be coerced to this type. Can be LONG, FLOAT, DOUBLE, STRING, ARRAY types, or COMPLEX types.|no, default is FLOAT|

### Nested field virtual column

The nested field virtual column is an optimized virtual column that can provide direct access into various paths of
a `COMPLEX<json>` column, including using their indexes.

> Syntax (all 3 of these virtual columns produce the same output):

Reviewer: ```suggestion
You can define a nested field virtual column with any of the following syntaxes.
The examples differ in how to access a specific value from the `COMPLEX<json>`.
All examples produce the same output value.
Reference a virtual column by its output name, such as `v0`, `v1`, or `v2`.
```

```json
    {
      "type": "nested-field",
      "columnName": "shipTo",
      "outputName": "v0",
      "expectedType": "STRING",
      "path": "$.phoneNumbers[1].number"
    }
```
```json
    {
      "type": "nested-field",
      "columnName": "shipTo",
      "outputName": "v1",
      "expectedType": "STRING",
      "path": ".phoneNumbers[1].number",
      "useJqSyntax": true
    }
```

```json
    {
      "type": "nested-field",
      "columnName": "shipTo",
      "outputName": "v2",
      "expectedType": "STRING",
      "pathParts": [
        {
          "type": "field",
          "field": "phoneNumbers"
        },
        {
          "type": "arrayElement",
          "index": 1
        },
        {
          "type": "field",
          "field": "number"
        }
      ]
    }
```

|property|description|required?|
|--------|-----------|---------|
|columnName|The name of the virtual column.|yes|
|outputName|The name of the virtual column.|yes|
|expectedType|The name of the virtual column.|yes|

> |pathParts|The name of the virtual column.|yes|

Reviewer: Update description?

Author: heh oops

|processFromRaw|If set to true, the virtual column will process the "raw" JSON data to extract values rather than using an optimized "literal" value selector. This option allows extracting non-literal values (such as nested JSON objects or arrays) as a `COMPLEX<json>` at the cost of much slower performance.|No, default false|
|path|'JSONPath' or 'jq' syntax path. One of `path` or `pathParts` must be set|no, if `pathParts` is defined|

> |useJqSyntax||no, default is false|

Reviewer: Missing description?

> #### Nested path part

Reviewer: ```suggestion
#### Nested path part
Specify `pathParts` as an array of objects that describe each component of the path to traverse. Each object can take the following properties:
```

Author: added :+1:

|property|description|required?|
|--------|-----------|---------|

> |type|Must be 'field' or 'arrayElement'|yes|

Reviewer: ```suggestion
|type|Must be `field` or `arrayElement`. Use `field` when accessing a specific field in a nested structure. Use `arrayElement` when accessing a specific integer position of an array.|yes|
```
Array position starts with 0?

|field|The name of the 'field' in a 'field' `type` path part|yes, if `type` is 'field'|
|index|The array element index if `type` is `arrayElement`|yes, if `type` is 'arrayElement'|

> This virtual column is used for the SQL operators `JSON_VALUE` (if `processFromRaw` is set to false) or `JSON_QUERY`

Reviewer: This paragraph might be more useful at the beginning of the section (before the syntax code blocks)

Author: moved, though it's not really necessary to care how SQL planner uses this native virtual column to implement its functions, so it could also be removed probably. Its just adding flavor about what this thing is for I guess.

Reviewer: ah true. the helpful part for the intro was more so on how you can use JSONPath or jq syntax to extract part of the json

> (if it is true), and accepts 'JSONPath' or 'jq' syntax string representations of paths, or a parsed

Reviewer: ```suggestion
(if `processFromRaw` is true), and accepts 'JSONPath' or 'jq' syntax string representations of paths, or a parsed
```

list of "path parts" in order to determine what should be selected from the column.

Type information for nested fields is absent at higher levels (it is contained within the segment, but not to segment
metadata queries or the SQL planner), so `expectedType` provides the context for how something is being used, e.g. an

> aggregators default type or an explicit cast, or, if using the 'RETURNING' syntax which explicitly specifies type.

Reviewer: ```suggestion
aggregators default type or an explicit cast, or, if using the 'RETURNING' syntax in `JSON_VALUE` to explicitly specify type.
```

Author: i removed this section and just expanded the information for `expectedType` to indicate that druid coerces values to the specified type if it is supplied.
